proposals/TP2-dynamically-generated-cacheable-xds-resources.md

```diff
@@ -0,0 +1,759 @@
+  that means that we also need two different versions of the CDS
+  resource, also distinguished by the same context parameter.  And then
+  we need two different versions of the RDS resource, since that needs
+  to refer to the CDS resource.  And then two different versions of the
+  LDS resource, which refers to the RDS resource.  This causes a
+  combinatorial explosion in the number of resources needed, and it adds
+  complexity to xDS servers, which need to construct the right variants
+  of every resource and make sure that they refer to each other using
+  the right names.
+- In the new xDS naming scheme, context parameters are exact-match-only.
+  This means that if a control plane wants to provide the same resource
+  both with and without a given parameter, it needs to publish two
+  versions of the resource, each with a different name, even though the
+  contents are the same, which can also cause unnecessarily poor cache
+  performance.  For example, in the "dynamic route selection" use-case,
+  let's say that every client uses two different dynamic selection
+  parameters, `env` (which can have one of the values `prod`, `canary`, or
+  `test`) and `version` (which can have one of the values `v1`, `v2`, or
+  `v3`).  Now let's say that there is a `RouteConfiguration` with one route
+  that should be selected via the parameter `env=prod` and another route that
+  should be selected via the parameter `version=v1`. This means that there
+  are only four variants of the `RouteConfiguration` resource (`{env!=prod,
+  version!=v1}`, `{env=prod, version!=v1}`, `{env!=prod, version=v1}`, and
+  `{env=prod, version=v1}`).  However, the exact-match semantics means
+  that there will have to be nine different versions of this resource,
+  one for each combination of values of the two parameters.
+
+### Related Proposals:
+* [xRFC TP1: new xDS naming scheme](TP1-xds-transport-next.md)
+
+## Proposal
+
+This document proposes an alternative approach.  We start with the
+observation that resource names are used in two places:
+
+- The **transport protocol** layer, which needs to identify the right
+  resource contents to send for a given resource name, often obtaining
+  those resource contents from a cache.
+- The **resource graph** used on an individual client, where there are a
+  set of data model resources that refer to each other by name.  For
+  example, a `RouteConfiguration` refers to individual `Cluster` resources
+  by name.
+
+The use-cases for dynamic resource selection share one important property
+that we can take advantage of.  When multiple variants of a given resource
+exist, any given client will only ever use one of those variants at a
+given time.  That means that the parameters that affect which variant
+of the resource is used are required by the transport protocol, but
+they are not required by the client's data model.  (For example, in the
+"sharding endpoints for scalability" use-case, different clients may see
```

Comment:
I thought this use case is out of context for this design?

Reply:
Good catch, this is a lingering comment from an earlier draft.  Removed.